light/beacon/sync_committee_update.go

```diff
@@ -0,0 +1,545 @@
+	broadcastFrequencyLimit = time.Millisecond * 200
+	advertiseDelay          = time.Second * 10
+	lastProcessedCount      = 4
+)
+
+// CommitteeRequest represents a request for fetching updates and committees at the given periods
+type CommitteeRequest struct {
+	UpdatePeriods, CommitteePeriods []uint64
+}
+
+// empty returns true if the request does not request anything
+func (req CommitteeRequest) empty() bool {
+	return req.UpdatePeriods == nil && req.CommitteePeriods == nil
+}
+
+// CommitteeReply is an answer to a CommitteeRequest, contains the updates and committees corresponding
+// to the period numbers in the request in the same order
+type CommitteeReply struct {
+	Updates    []LightClientUpdate
+	Committees [][]byte
+}
+
+// sctClient represents a peer that SyncCommitteeTracker sends signed heads and sync committee advertisements to
+type sctClient interface {
+	SendSignedHeads(heads []SignedHead)
+	SendUpdateInfo(updateInfo *UpdateInfo)
+}
+
+// sctServer represents a peer that SyncCommitteeTracker can request sync committee update proofs from
+type sctServer interface {
+	GetBestCommitteeProofs(ctx context.Context, req CommitteeRequest) (CommitteeReply, error)
+	CanRequest(updateCount, committeeCount int) bool
+	WrongReply(description string)
+}
+
+// UpdateInfo contains scores for an advertised update chain. Note that the most recent updates are always
+// advertised but earliest ones might not because of length limitation.
+type UpdateInfo struct {
+	AfterLastPeriod uint64
+	Scores          UpdateScores
+}
+
+// UpdateScore allows the comparison between updates at the same period in order to find the best
+// update chain that provides the strongest proof of being canonical.
+type UpdateScore struct {
+	signerCount, subPeriodIndex uint32
+	finalizedHeader             bool // update is considered finalized if has finalized header and 2/3 signatures
+}
+
+type UpdateScores []UpdateScore
+
+// isFinalized returns true if the update has a header signed by at least 2/3 of the committee,
+// referring to a finalized header that refers to the next sync committee. This condition is a
+// close approximation of the actual finality condition that can only be verified by full beacon nodes.
+func (u *UpdateScore) isFinalized() bool {
+	return u.finalizedHeader && u.signerCount > 341
+}
+
+// reorgRiskPenalty returns a modifier that approximates the risk of a non-finalized update header being reorged.
+// It should be subtracted from the signer participation count when comparing non-finaized updates.
+// This risk is assumed to be dropping exponentially as the update header is further away from the beginning of the period.
+func (u *UpdateScore) reorgRiskPenalty() uint32 {
+	return uint32(math.Pow(2, 10-float64(u.subPeriodIndex)/32))
+}
+
+// betterThan returns true if update u is considered better than w.
+func (u *UpdateScore) betterThan(w UpdateScore) bool {
+	uFinalized, wFinalized := u.isFinalized(), w.isFinalized()
```

Comment:
From the spec, the comparison function is also defined, https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/light-client/sync-protocol.md#is_better_update

I feel like we should strictly follow the spec just re-implement it in golang.

Reply:
I thought about it a bit more, also discussed it with @wemeetagain and we agreed that we should try to change the spec. I opened a PR against the specs repo where I also explained the reasons:
https://github.com/ethereum/consensus-specs/pull/3128/
(it is the same algorithm as implemented in this PR)